import grpc
from concurrent import futures
import threading
import queue
import logging
import time
from datetime import datetime
import random

import service_pb2 as pb2
import service_pb2_grpc as pb2_grpc


# === –°–ü–ò–°–û–ö –ö–û–ú–ê–ù–î ===
SERVER_COMMANDS = {
    '!–≤—Ä–µ–º—è': lambda: f"üïê –°–µ–π—á–∞—Å: {datetime.now().strftime('%d.%m.%Y %H:%M:%S')}",
    '!–¥–∞—Ç–∞': lambda: f"üìÖ –°–µ–≥–æ–¥–Ω—è: {datetime.now().strftime('%d.%m.%Y')}",
    '!—á–∞—Å—ã': lambda: f"‚è∞ –í—Ä–µ–º—è: {datetime.now().strftime('%H:%M')}",
    '!—Ä–∞–Ω–¥–æ–º': lambda: f"üé≤ –°–ª—É—á–∞–π–Ω–æ–µ —á–∏—Å–ª–æ: {random.randint(1, 100)}",
    '!–º–æ–Ω–µ—Ç–∫–∞': lambda: f"ü™ô –ú–æ–Ω–µ—Ç–∫–∞: {'–û—Ä—ë–ª!' if random.choice([True, False]) else '–†–µ—à–∫–∞!'}",
    '!–∫—É–±–∏–∫': lambda: f"üé∂ –ö—É–±–∏–∫: {random.randint(1, 6)}",
    '!–ø–æ–º–æ—â—å': lambda: (
        "üìö –î–æ—Å—Ç—É–ø–Ω—ã–µ –∫–æ–º–∞–Ω–¥—ã:\n"
        "  !–≤—Ä–µ–º—è ‚Äî —Ç–µ–∫—É—â–µ–µ –≤—Ä–µ–º—è\n"
        "  !–¥–∞—Ç–∞ ‚Äî —Ç–µ–∫—É—â–∞—è –¥–∞—Ç–∞\n"
        "  !—á–∞—Å—ã ‚Äî —á–∞—Å—ã –∏ –º–∏–Ω—É—Ç—ã\n"
        "  !—Ä–∞–Ω–¥–æ–º ‚Äî —á–∏—Å–ª–æ –æ—Ç 1 –¥–æ 100\n"
        "  !–º–æ–Ω–µ—Ç–∫–∞ ‚Äî –ø–æ–¥–±—Ä–æ—Å–∏—Ç—å –º–æ–Ω–µ—Ç–∫—É\n"
        "  !–∫—É–±–∏–∫ ‚Äî –±—Ä–æ—Å–∏—Ç—å –∫—É–±–∏–∫ (1-6)\n"
        "  !–ø–æ–º–æ—â—å ‚Äî —ç—Ç–æ—Ç —Å–ø–∏—Å–æ–∫\n"
        "  !—Ü–≤–µ—Ç ‚Äî —Å–ª—É—á–∞–π–Ω—ã–π —Ü–≤–µ—Ç\n"
        "  !—Ñ–∞–∫—Ç ‚Äî —Å–ª—É—á–∞–π–Ω—ã–π —Ñ–∞–∫—Ç"
    ),
    '!—Ü–≤–µ—Ç': lambda: f"üé® –¶–≤–µ—Ç: #{random.randint(0, 0xFFFFFF):06X}",
    '!—Ñ–∞–∫—Ç': lambda: random.choice([
        "üß† –ú–æ–∑–≥ —á–µ–ª–æ–≤–µ–∫–∞ –≥–µ–Ω–µ—Ä–∏—Ä—É–µ—Ç –æ–∫–æ–ª–æ 70 000 –º—ã—Å–ª–µ–π –≤ –¥–µ–Ω—å",
        "üêô –£ –æ—Å—å–º–∏–Ω–æ–≥–∞ —Ç—Ä–∏ —Å–µ—Ä–¥—Ü–∞",
        "üçØ –ú—ë–¥ –Ω–∏–∫–æ–≥–¥–∞ –Ω–µ –ø–æ—Ä—Ç–∏—Ç—Å—è (–∞—Ä—Ö–µ–æ–ª–æ–≥–∏ –Ω–∞—Ö–æ–¥–∏–ª–∏ —Å—ä–µ–¥–æ–±–Ω—ã–π –º—ë–¥ –≤ –≥—Ä–æ–±–Ω–∏—Ü–∞—Ö)",
        "üå°Ô∏è –°–∞–º–∞—è –≤—ã—Å–æ–∫–∞—è —Ç–µ–º–ø–µ—Ä–∞—Ç—É—Ä–∞ –Ω–∞ –ó–µ–º–ª–µ: +56.7¬∞C (–î–æ–ª–∏–Ω–∞ –°–º–µ—Ä—Ç–∏, 1913)",
        "üöÄ –°–≤–µ—Ç –æ—Ç –°–æ–ª–Ω—Ü–∞ –¥–æ –ó–µ–º–ª–∏ –∏–¥—ë—Ç 8 –º–∏–Ω—É—Ç 20 —Å–µ–∫—É–Ω–¥",
        "üêå –£–ª–∏—Ç–∫–∞ –º–æ–∂–µ—Ç —Å–ø–∞—Ç—å 3 –≥–æ–¥–∞",
        "üçï –í –º–∏—Ä–µ –µ–∂–µ–¥–Ω–µ–≤–Ω–æ —Å—ä–µ–¥–∞—é—Ç –æ–∫–æ–ª–æ 350 –∫—É—Å–∫–æ–≤ –ø–∏—Ü—Ü—ã –≤ —Å–µ–∫—É–Ω–¥—É",
        "üì± –ü–µ—Ä–≤—ã–π —Ç–µ–ª–µ—Ñ–æ–Ω–Ω—ã–π –∑–≤–æ–Ω–æ–∫ –±—ã–ª —Å–¥–µ–ª–∞–Ω –≤ 1876 –≥–æ–¥—É",
    ]),
}


class ChatService(pb2_grpc.ChatServiceServicer):
    def __init__(self):
        self.clients: dict[str, dict] = {}  # {username: {queue, emoji}}
        self.lock = threading.Lock()
        logging.info("üí¨ ChatService initialized with commands")

    def ChatStream(self, request_iterator, context):
        username = None
        emoji = "üòÄ"  # –°–º–∞–π–ª–∏–∫ –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é
        outgoing_queue: queue.Queue = queue.Queue()

        def read_incoming():
            nonlocal username, emoji
            try:
                for request in request_iterator:
                    if not username:
                        # –ü–µ—Ä–≤–æ–µ —Å–æ–æ–±—â–µ–Ω–∏–µ: —Ä–µ–≥–∏—Å—Ç—Ä–∏—Ä—É–µ–º –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
                        username = request.username.strip() or f"User_{id(context) % 1000}"
                        emoji = request.emoji.strip() if request.emoji else "üòÄ"
                        
                        with self.lock:
                            self.clients[username] = {"queue": outgoing_queue, "emoji": emoji}
                        
                        logging.info(f"üü¢ –ü–æ–¥–∫–ª—é—á–∏–ª—Å—è: {username} {emoji}")
                        self._broadcast("SERVER", f"{username} {emoji} –∑–∞—à–µ–ª –≤ —á–∞—Ç", exclude=username)
                    else:
                        # –ü—Ä–æ–≤–µ—Ä—è–µ–º –Ω–∞ –∫–æ–º–∞–Ω–¥—ã
                        text = request.text.strip()
                        if text.startswith('!'):
                            self._handle_command(text, username)
                        else:
                            logging.info(f"üì® {username} {emoji}: {text}")
                            self._broadcast(username, text, exclude=username, emoji=emoji)
                
            except grpc.RpcError as e:
                logging.warning(f"‚ö†Ô∏è –û—à–∏–±–∫–∞ —á—Ç–µ–Ω–∏—è —É {username}: {e.code()}")
            except Exception as e:
                logging.error(f"‚ùå –ö—Ä–∏—Ç–∏—á–µ—Å–∫–∞—è –æ—à–∏–±–∫–∞ —É {username}: {e}")
            finally:
                if username:
                    logging.info(f"üî¥ –û—Ç–∫–ª—é—á–∏–ª—Å—è: {username}")
                    self._broadcast("SERVER", f"{username} {emoji} –ø–æ–∫–∏–Ω—É–ª —á–∞—Ç", exclude=username)
                    with self.lock:
                        self.clients.pop(username, None)

        reader_thread = threading.Thread(target=read_incoming, daemon=True)
        reader_thread.start()

        try:
            while context.is_active():
                try:
                    msg = outgoing_queue.get(timeout=1.0)
                    yield msg
                except queue.Empty:
                    continue
        except grpc.RpcError as e:
            logging.warning(f"‚ö†Ô∏è –û—à–∏–±–∫–∞ –∑–∞–ø–∏—Å–∏: {e.code()}")
        finally:
            reader_thread.join(timeout=2.0)

    def _handle_command(self, text: str, username: str):
        """–û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –∫–æ–º–∞–Ω–¥—ã —Å–µ—Ä–≤–µ—Ä–∞"""
        command = text.lower().split()[0]  # –ë–µ—Ä—ë–º –ø–µ—Ä–≤–æ–µ —Å–ª–æ–≤–æ (–∫–æ–º–∞–Ω–¥—É)
        
        if command in SERVER_COMMANDS:
            try:
                result = SERVER_COMMANDS[command]()
                # –û—Ç–ø—Ä–∞–≤–ª—è–µ–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç –í–°–ï–ú (–≤–∫–ª—é—á–∞—è –æ—Ç–ø—Ä–∞–≤–∏—Ç–µ–ª—è)
                self._broadcast("SERVER", f"{username}: {text}\n{result}", exclude=None)
            except Exception as e:
                self._broadcast("SERVER", f"‚ö†Ô∏è –û—à–∏–±–∫–∞ –∫–æ–º–∞–Ω–¥—ã: {e}", exclude=None)
        else:
            # –ù–µ–∏–∑–≤–µ—Å—Ç–Ω–∞—è –∫–æ–º–∞–Ω–¥–∞ ‚Äî –ø–æ–∫–∞–∑—ã–≤–∞–µ–º –ø–æ–¥—Å–∫–∞–∑–∫—É
            self._broadcast("SERVER", f"‚ùì –ù–µ–∏–∑–≤–µ—Å—Ç–Ω–∞—è –∫–æ–º–∞–Ω–¥–∞ '{command}'. –í–≤–µ–¥–∏—Ç–µ !–ø–æ–º–æ—â—å", exclude=username)

    def _broadcast(self, sender: str, text: str, exclude: str = None, emoji: str = ""):
        """–û—Ç–ø—Ä–∞–≤–ª—è–µ—Ç —Å–æ–æ–±—â–µ–Ω–∏–µ –≤—Å–µ–º –∫–ª–∏–µ–Ω—Ç–∞–º, –∫—Ä–æ–º–µ exclude"""
        msg = pb2.ChatMessage(
            username=sender,
            text=text,
            emoji=emoji
        )
        with self.lock:
            for user_name, data in list(self.clients.items()):
                if exclude and user_name == exclude:
                    continue
                try:
                    data["queue"].put(msg)
                except Exception:
                    pass


def serve():
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=100),
        options=[
            ('grpc.keepalive_time_ms', 60000),
            ('grpc.keepalive_timeout_ms', 10000),
            ('grpc.keepalive_permit_without_calls', 1),
            ('grpc.http2.max_pings_without_data', 2),  # –†–∞–∑—Ä–µ—à–∞–µ–º 2 –ø–∏–Ω–≥–∞ –±–µ–∑ –¥–∞–Ω–Ω—ã—Ö
            ('grpc.http2.min_ping_interval_without_data_ms', 30000), # –ú–∏–Ω–∏–º—É–º 30 —Å–µ–∫ –º–µ–∂–¥—É –ø–∏–Ω–≥–∞–º–∏
        ]
    )
    pb2_grpc.add_ChatServiceServicer_to_server(ChatService(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    logging.info("‚úÖ –°–µ—Ä–≤–µ—Ä —á–∞—Ç–∞ –∑–∞–ø—É—â–µ–Ω –Ω–∞ [::]:50051")
    logging.info("üåê –î–æ—Å—Ç—É–ø–µ–Ω –∏–∑ –ª–æ–∫–∞–ª—å–Ω–æ–π —Å–µ—Ç–∏ –∏ Tailscale")
    logging.info("üìö –ö–æ–º–∞–Ω–¥—ã: !–≤—Ä–µ–º—è, !–¥–∞—Ç–∞, !—Ä–∞–Ω–¥–æ–º, !–º–æ–Ω–µ—Ç–∫–∞, !–∫—É–±–∏–∫, !–ø–æ–º–æ—â—å, !—Ü–≤–µ—Ç, !—Ñ–∞–∫—Ç")

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        logging.info("üõë –ü–æ–ª—É—á–µ–Ω —Å–∏–≥–Ω–∞–ª –æ—Å—Ç–∞–Ω–æ–≤–∫–∏, –∑–∞–≤–µ—Ä—à–∞–µ–º —Ä–∞–±–æ—Ç—É...")
        server.stop(0)


if __name__ == '__main__':
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s [%(levelname)s] %(message)s',
        datefmt='%H:%M:%S'
    )
    serve()